src/Sphere.jl

```diff
@@ -156,3 +156,75 @@ function normal_tvector_distribution(S::Sphere, x, σ)
     d = Distributions.MvNormal(zero(x), σ)
     return ProjectedFVectorDistribution(TangentBundleFibers(S), x, d, project_vector!, x)
 end
+
+struct GeodesicInterpolationMethod <: AbstractMethod end
+
+"""
+    mean!(
+        S::Sphere,
+        y,
+        x::AbstractVector,
+        w::AbstractWeights,
+        method::GeodesicInterpolationMethod;
+        shuffle_rng=nothing,
+        kwargs...,
+    )
+
+Compute the Riemannian mean (center of mass) of `x`. The algorithm is adapted
+from the below paper to handle the weighted mean. It uses repeated weighted
+geodesic interpolation and converges to the Riemannian mean when all `x` are
+within a distance of π/2 of the mean.
+
+If `shuffle_rng` is provided, it is used to shuffle the order in which the
+points are considered for computing the mean.
+
+For more information on the geodesic interpolation method, see the below paper:
+> Salehian H et al: An efficient recursive estimator of the
+> Fréchet mean on a hypersphere with applications to Medical Image Analysis,
+> Mathematical Foundations of Computational Anatomy (2015).
+"""
+function mean!(
```

Comment:
Would the geodesic interpolation maybe also something one could provide for all manifolds? Or is it really only useful here? It seems to do something reasonable on manifolds in general. but I can't follow this completely – it seems it is then only used to initialise the gradient descent and is not a mean itself? Then the name mean (with GeodesicInterpolationMethod) is misleading?

Reply:
The geodesic interpolation algorithm is based on the geometric interpretation of incremental mean calculation as a weighted interpolation along the geodesic between the existing mean and the new datapoint. I don't know whether this should converge to the mean on any manifold.

It's general enough that we could leave it as a general method that others can test on their manifold. On the sphere, the paper only proved convergence to the mean when all the datapoints are within a geodesic ball of radius pi/2 from the mean. Hence the default on the `Sphere` is to compute the mean this way, then check the condition. If it doesn't apply, the estimate is used to initialize gradient descent because from manual checking, even if the condition is violated, the mean tends to be close to the mean computed by gradient descent. And the algorithm tends to be quite a bit faster than the gradient descent approach.

Reply:
If I understand the paper correctly, it's actually an estimator for the mean of the distribution from which the points `x` were drawn. They prove that on spheres it converges and is unbiased.

Reply:
I think though it is worth noting that it only converges to the same mean as the gradient descent algorithm asymptotically with increasing sample size. For low sample size, it may give a different answer. Perhaps even if the condition is satisfied, it should always be used to initialize the gradient descent approach?

Reply:
There is for example also no convergence theory for the Gradient method on Hadamard manifolds or for the CPPA approach on the Sphere (a reason why I would like to introduce sub gradient).

Using it as an initialisation to a different method that aims to solve the same problem was, what got me wondering first, I think. Is it good enough in approximation? Then we don't need to continue with grad desc. Is it not good enough? Then we maybe should not call it mean?

Reply:
Well it's good enough under conditions, but those conditions have only been worked out for a single manifold. When called with that method explicitly requested, we trust that the user knows what they're doing, and I think that's the right way to go. Only in the default case where the user doesn't care what algorithm is used would we maybe fallback. So I don't think the method name is misleading as long as the documentation is clear.

Reply:
Ok. But then let's not directly use it as a start for the gradient descent.

And of course, document it clearly.

Reply:
> Ok. But then let's not directly use it as a start for the gradient descent.

In the general case yes (i.e. for `mean!(::Manifold, ::Any, ::AbstractVector, ::AbstractWeights, ::GeodesicInterpolationMethod; kwargs...)` , but for `mean!(::Sphere, ::Any, ::AbstractVector, ::AbstractWeights; kwargs...)` (i.e. the default for `Sphere`) where we know the conditions under which the estimate is good, it's important to fall back to gradient descent if the conditions are violated.